src/AutoRest.CSharp/Common/Output/Models/Types/SignatureType.cs

```diff
@@ -0,0 +1,254 @@
+    /// </summary>
+    internal class SignatureType
+    {
+        private readonly string _defaultNamespace;
+        private readonly string _defaultName;
+        private readonly SignatureType? _customization;
+        private readonly SignatureType? _baselineContract;
+        private readonly MethodChangeset? _methodChangeset;
+
+        // Missing means the method with the same name is missing from the current contract
+        // Updated means the method with the same name is updated in the current contract, and the list contains the previous method and current methods including overload ones
+        private record MethodChangeset(IReadOnlyList<MethodSignature> Missing, IReadOnlyList<(List<MethodSignature> Current, MethodSignature Previous)> Updated) { }
+
+        public SignatureType(IReadOnlyList<MethodSignature> methods, SourceInputModel? sourceInputModel, string defaultNamespace, string defaultName)
+        {
+            Methods = methods;
+            _defaultNamespace = defaultNamespace;
+            _defaultName = defaultName;
+            if (sourceInputModel is not null)
+            {
+                _customization = new SignatureType(PopulateMethodsFromCompilation(sourceInputModel?.Customization), null, defaultNamespace, defaultName);
+                _baselineContract = new SignatureType(PopulateMethodsFromCompilation(sourceInputModel?.PreviousContract), null, defaultNamespace, defaultName);
+                _methodChangeset ??= CompareMethods(Methods.Union(_customization?.Methods ?? Array.Empty<MethodSignature>(), MethodSignature.ParameterAndReturnTypeEqualityComparer), _baselineContract?.Methods);
+            }
+        }
+
+        private IReadOnlyList<OverloadMethodSignature>? _overloadMethods;
+        public IReadOnlyList<OverloadMethodSignature> OverloadMethods => _overloadMethods ??= EnsureOverloadMethods();
+
+        private IReadOnlyList<OverloadMethodSignature> EnsureOverloadMethods()
+        {
+            var overloadMethods = new List<OverloadMethodSignature>();
+            var updated = _methodChangeset?.Updated;
+            if (updated is null)
+            {
+                return Array.Empty<OverloadMethodSignature>();
+            }
+
+            foreach (var (current, previous) in updated)
+            {
+                if (TryGetPreviousMethodWithLessOptionalParameters(current, previous, out var currentMethodToCall, out var missingParameters))
+                {
+                    overloadMethods.Add(new OverloadMethodSignature(currentMethodToCall, previous.DisableOptionalParameters(), missingParameters, previous.Description, true));
+                }
+            }
+            return overloadMethods;
+        }
+
+        private HashSet<MethodSignature>? _methodsToSkip;
+        public HashSet<MethodSignature> MethodsToSkip => _methodsToSkip ??= EnsureMethodsToSkip();
```

Comment:
maybe `ImmutableHashSet` is better here since we do not have a `IReadOnlySet`

Reply:
Oh I was wrong - we have a `IReadOnlySet<T>` interface and here we should use `IReadOnlySet<T>` as the type

Reply:
Updated